File: Python/24.Swap Nodes in Pairs.py
#把链表中的两个两个节点互相交换(不能直接换数字)
#如1->2->3->4变为2->1->4->3
class ListNode(object):
     def __init__(self, x):
         self.val = x
         self.next = None

class Solution(object):
    def swapPairs(self, head):
        nHead = ListNode(0)
        nHead.next = head
        p = nHead
        q = nHead.next
        while q:
            t = q.next
            if t:
                p.next = t
                q.next = t.next
                t.next = q
            p = p.next.next
            q = q.next
        return nHead.next

l1 = ListNode(1)
l2 = ListNode(2)
l3 = ListNode(3)
l4 = ListNode(4)
l5 = ListNode(5)
l1.next = l2
l2.next = l3
l3.next = l4
l4.next = l5
s = Solution()
p = s.swapPairs(l1)
while p:
    print(p.val)
    p = p.next

File: C++/Linked List/24.Swap Nodes in Pairs.cpp
/*
将链表中节点两个两个互换。
1->2->3->4返回2->1->4->3
只能用常数的空间，不能动链表中的值。
*/

#include <iostream>
#include <vector>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* swapPairs(ListNode* head) {
		if(!head)	return NULL;	
        ListNode nHead(0);
        nHead.next = head;
        ListNode* pre = &nHead, cur = head, past;
        while(cur && (past=cur->next))
        {
        	pre->next = past;
        	pre = cur;
        	cur = past->next;
        	past->next = pre;
        	pre->next = cur;
        }
        return nHead.next;
};
int main()
{
	Solution s;
	ListNode head(2);
	ListNode t1(5);
	ListNode t2(10);
	ListNode t3(3);
	ListNode t4(6);
	ListNode t5(13);
	head.next = &t1;
	t1.next = &t2;
	t2.next = &t3;
	t3.next = &t4;
	t4.next = &t5;
 	ListNode* h = s.swapPairs(&head);
 	while(h)
 	{
 		cout<<h->val<<endl;
 		h = h->next;
 	}
}

